subsys/bluetooth/controller/ll_sw/ull_iso.c

```diff
@@ -625,24 +872,349 @@ uint8_t ll_read_iso_link_quality(uint16_t  handle,
 #endif /* CONFIG_BT_CTLR_SYNC_ISO || CONFIG_BT_CTLR_CONN_ISO */
 
 #if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
+static isoal_status_t ll_iso_test_pdu_release(struct node_tx_iso *node_tx,
+					      const uint16_t handle,
+					      const isoal_status_t status)
+{
+	/* Release back to memory pool */
+	ll_iso_link_tx_release(node_tx->link);
+	ll_iso_tx_mem_release(node_tx);
+
+	return ISOAL_STATUS_OK;
+}
+
+#if defined(CONFIG_BT_CTLR_CONN_ISO)
+static void ll_iso_transmit_test_send_sdu(uint16_t handle, uint32_t ticks_at_expire)
+{
+	isoal_source_handle_t source_handle;
+	struct isoal_sdu_tx sdu;
+	isoal_status_t err;
+	uint8_t tx_buffer[ISO_TEST_TX_BUFFER_SIZE];
+	uint16_t remaining_tx;
+	uint32_t sdu_counter;
+
+	if (IS_CIS_HANDLE(handle)) {
+		struct isoal_pdu_production *pdu_production;
+		struct ll_conn_iso_stream *cis;
+		struct ll_conn_iso_group *cig;
+		struct isoal_source *source;
+		uint32_t rand_max_sdu;
+		uint8_t rand_8;
+
+		cis = ll_iso_stream_connected_get(handle);
+		LL_ASSERT(cis);
+
+		if (!cis->hdr.test_mode.tx_enabled) {
+			/* Transmit Test Mode not enabled */
+			return;
+		}
+
+		cig = cis->group;
+		source_handle = cis->hdr.datapath_in->source_hdl;
+
+		switch (cis->hdr.test_mode.tx_payload_type) {
+		case BT_HCI_ISO_TEST_ZERO_SIZE_SDU:
+			remaining_tx = 0;
+			break;
+
+		case BT_HCI_ISO_TEST_VARIABLE_SIZE_SDU:
+			/* Randomize the length [4..p_max_sdu] */
+			lll_rand_get(&rand_8, sizeof(rand_8));
+			rand_max_sdu = rand_8 * (cis->p_max_sdu - 4);
+			remaining_tx = 4 + (rand_max_sdu >> 8);
+			break;
+
+		case BT_HCI_ISO_TEST_MAX_SIZE_SDU:
+			LL_ASSERT(cis->p_max_sdu > 4);
+			remaining_tx = cis->p_max_sdu;
+			break;
+
+		default:
+			LL_ASSERT(0);
+			return;
+		}
+
+		if (remaining_tx > ISO_TEST_TX_BUFFER_SIZE) {
+			sdu.sdu_state = BT_ISO_START;
+		} else {
+			sdu.sdu_state = BT_ISO_SINGLE;
+		}
+
+		/* Configure SDU similarly to one delivered via HCI */
+		sdu.dbuf = tx_buffer;
+		sdu.cig_ref_point = cig->cig_ref_point;
+		sdu.target_event = cis->lll.event_count + (cis->lll.tx.flush_timeout > 1 ? 0 : 1);
```

Comment:
General comment: IIRC we generally add `U` to unsigned values in the controller, e.g.
```suggestion
		sdu.target_event = cis->lll.event_count + (cis->lll.tx.flush_timeout > 1U ? 0U : 1U);
```

Reply:
This applies many places in this PR

Reply:
Checking ull.c and ull_conn.c, it looks like about 50% of constants have the `U` ;-). Anyways, I will add it.